php/base-stemmer.php

```diff
@@ -0,0 +1,524 @@
+                $diff = $this->currentCharCodeAt($c - 1 - $common) - self::charCodeAt($w[0],$i2);
+                if ($diff != 0){
+                    break;
+                }
+                $common++;
+            }
+            if ($diff < 0) {
+                $j = $k;
+                $common_j = $common;
+            }
+            else {
+                $i = $k;
+                $common_i = $common;
+            }
+            if ($j - $i <= 1) {
+                if ( $i > 0 || $j === $i || $first_key_inspected ){
+                    break;
+                }
+                $first_key_inspected = true;
+            }
+        }
+        do {
+            $w = $v[$i];
+            $w0length = self::lengthOf($w[0]);
+            if ( $common_i >= $w0length ){
+                $this->cursor = $c - $w0length;
+                if ( count($w) < 4){
+                    return $w[2];
+                }
+                $res = call_user_func( [ $this, $w[3] ] );
+                $this->cursor = $c - $w0length;
+                if ($res){
+                    return $w[2];
+                }
+            }
+            $i = $w[1];
+        } while ($i >= 0);
+        return 0;
+    }
+
+
+    /**
+     * to replace chars between $c_bra and $c_ket in $this->current by the chars in $s.
+     */
+    private function replace_s( int $c_bra, int $c_ket, string $s ):int {
+        $slength = self::lengthOf($s);
+        $adjustment = $slength - ($c_ket - $c_bra);
+        $this->current = $this->currentSlice(0,$c_bra) . $s . $this->currentSlice($c_ket);
+        $this->limit += $adjustment;
+        if ( $this->cursor >= $c_ket){
+            $this->cursor += $adjustment;
+        }
+        else if ( $this->cursor > $c_bra){
+            $this->cursor = $c_bra;
+        }
+        return $adjustment;
+    }
+
+
+    private function slice_check():bool {
+        return ! (
+            $this->bra < 0 ||
+            $this->bra > $this->ket ||
+            $this->ket > $this->limit ||
+            $this->limit > $this->currentLength()
+        );
+    }
+
+
+    protected function slice_from( string $s ):bool {
+        if( $this->slice_check() ){
+            $this->replace_s($this->bra, $this->ket, $s );
+            return true;
+        }
+        return false;
+    }
+
+
+    protected function slice_del():bool {
+        return $this->slice_from('');
+    }
+
+
+    protected function insert( int $c_bra, int $c_ket, string $s ):void {
+        $adjustment = $this->replace_s($c_bra, $c_ket, $s);
+        $c_bra <= $this->bra and $this->bra += $adjustment;
+        $c_bra <= $this->ket and $this->ket += $adjustment;
+    }
+
+
+    protected function slice_to():string {
+        $result = '';
+        if( $this->slice_check() ){
+            $result = $this->currentSlice($this->bra, $this->ket);
+        }
+        return $result;
```

Comment:
Note: this is buggy in Javascript (#242) - the slice can be empty so `slice_to` can validly return an empty string and so we can't have an empty string being returned as the way to signal an error.

Probably slice_check() should just assert, since this can only happen for a bug in the algorithm, Snowball compiler or runtime, and an out of range value won't result in a buffer overrun like it can in C.